README.md
# Chat of extended interaction

Read this in another language | [English](/README.md) | [Русский](/docs/ru/README.md)
|---|---|---|

## Quick Links

[Changelog](CHANGELOG.md) | [Contributing](CONTRIBUTING.md)
|---|---|

## Contents

* [Overview](#overview)
* [FAQ](#faq)
    * [How to open the chat?](#open-chat)
    * [Hotkeys](#hotkeys)
    * [Where to get the mod?](#get-mod)
    * [Commands](#commands)
* [Dependencies](#dependencies)
    * [Required](#required)
    * [Optional](#optional)
* [Issues](#issue)
* [Features](#feature)
* [Future plans](#Future-plans)
* [Special thanks](special-thanks)
* [Installing](#installing)
* [License](#license)

## Overview

Add gui of chat, new commands, new types of chat, new interactions.
Top left appear the new gui with drop-down with filters and other functions, that sends your message from the text field to the chat and can simplify the interaction of other elements.

## FAQ

### <a name="open-chat"></a> How to open the chat?

* variant 1: When your character appears.
* variant 2: Press any hothey of the mod.
* variant 3: Write to chat "/toggle-chat".

### Hotkeys

| Desription | Hotkey (Default) |
| -------- | ---- |
| Send a message   | Y   |
| Send a localised message   | SHIFT + Y   |
| Recover a last message   | CONTROL + Y   |
| Select a recipient for the private chat  | Middle mouse button  |
| Select a recipient for the faction chat  | SHIFT + Middle mouse button   |

### <a name="get-mod"></a> Where to get the mod?

You can either download a zip archive from [mods.factorio.com][homepage] or the [GitLab repository](https://gitlab.com/ZwerOxotnik/secondary-chat/tags).

### Commands

* /a \<message\> or /allied-send \<message\> - Sends a message to the allies.
* /l \<message\> or /local-send \<message\> - Sends a message to the nearest players.
* /surface-send \<message\> - Sends a message all players on your surface.
* /admins-send \<message\> - Sends a message to admins.
* /toggle-chat [\<all/faction/allied/local/surface/admins/drop-down\>] - Disabled or re-enables your own parts of additional chat.

## Dependencies

### Required

* [Event listener](https://mods.factorio.com/mod/event-listener)

### Optional

* [Color picker](https://forums.factorio.com/viewtopic.php?f=97&t=30657)

## <a name="issue"></a> Found an Issue?

Please report any issues or a mistake in the documentation, you can help us by
[submitting an issue](https://gitlab.com/ZwerOxotnik/secondary-chat/issues) to our GitLab Repository or on [mods.factorio.com](https://mods.factorio.com/mod/secondary-chat/discussion).

## <a name="feature"></a> Want a Feature?

You can *request* a new feature by [submitting an issue](https://gitlab.com/ZwerOxotnik/secondary-chat/issues) to our GitHub
Repository or on [mods.factorio.com](https://mods.factorio.com/mod/secondary-chat/discussion).

## Future plans

* Remake chat
* Completely alter and add new features
* Integration with other mods and etc.
* Provides API to let other mods accessing
* Extend settings for players, admins
* Add and change new events
* Change chat structure
* Change advanced settings
* Add new localizations
* Add Polish localization
* Add blacklist for players
* Add custom chat logs
* Add custom quick message
* Add settings for chats
* Add profiles
* Add export/import of settings

## Special thanks

* **MeteorSbor** - tester
* **HAKER PLAY** - translator
* **anonymous#1** - translator

## Installing

If you have downloaded a zip archive:

* simply place it in your mods directory.

For more information, see [Installing Mods on the Factorio wiki](https://wiki.factorio.com/index.php?title=Installing_Mods).

If you have downloaded the source archive (GitLab):

* copy the mod directory into your factorio mods directory
* rename the mod directory to secondary-chat_*versionnumber*, where *versionnumber* is the version of the mod that you've downloaded (e.g., 1.21.0)

## License

This project is copyright © 2017-2019 ZwerOxotnik \<zweroxotnik@gmail.com\>.

Use of the source code included here is governed by the European Union Public License v. 1.2 only. See the [LICENCE](/LICENCE) file for details.

[homepage]: http://mods.factorio.com/mod/secondary-chat
[Factorio]: https://factorio.com/


CHANGELOG.md
# Changelog

## 2019-04-08

### [v1.21.0][v1.21.0]

- Extended the mod interface
- Bugfixes
- Improved compatibility
- Code refactoring
- Changed the mod interface

## 2019-04-07

### [v1.20.1][v1.20.1]

* Improved compatibility

## 2019-03-31

### [v1.20.0][v1.20.0]

* Fixed sending messages by Enter
* Fixed interactions with filters
* Fixed selecting a recipient by hotkey for the private/faction chat when there is no the chat interface
* Fixed filters for selecting a recipient by hotkey for the faction chat
* Fixed localised messages of the local chat
* Fixed extended settings
* Changed style of the extended interface of settings
* The interface is turned off together with the extended interface of settings.
* Non-interacting stuff are hidden.

## 2019-03-09

### [v1.19.2][v1.19.2]

* Fixed loading a save (but unstable deleting and creating commands in the mod now)

## 2019-03-05

### [v1.19.1][v1.19.1]

* Another dependencies

## 2018-03-03

### [v1.19.0][v1.19.0]

* Added: module «[Event listener](https://gitlab.com/ZwerOxotnik/event-listener)»
* Changed: output of local chat
* Changed: output of notice
* Bugfixes

### [v1.18.0][v1.18.0]

* Updated for Factorio 0.17
* Added: auto focus after send message
* Added: events "on_send_message", "on_toggle", "on_pre_remove_mod", "on_update_gui_chat", "on_create_gui_chat", "on_pre_delete_gui_chat", "on_hide_gui_chat", "on_unhide_gui_chat", "on_create_button", "on_pre_delete_button", "on_add_chat", "on_delete_chat", "on_changed_parameter_setting", "on_update_gui_list_settings", "on_update_gui_container_settings" of the mod
* Added: remote interface "secondary-chat"
* Added: admins chat
* Added: /admins-send \<message\> - Sends a message to admins.
* Added: private chat
* Added: between factions chat
* Added: surface chat
* Added: /surface-send \<message\> - Sends a message all players on your surface.
* Added: filters for chats
* Added: CHANGELOG.md and README.md and CONTRIBUTING.md
* Added: shell scripts for interactions with repository on GitLab
* Added: response on events of scenario PvP
* Added: hotkey for send a message (Default: Y)
* Added: hotkey for send a localised message (Default: SHIFT + Y)
* Added: hotkey for recover a last message (Default: CONTROL + Y)
* Added: hotkey for select a recipient for the private chat (Default: Middle mouse button), point the cursor at an entity
* Added: hotkey for select a recipient for the faction chat (Default: SHIFT + Middle mouse button), point the cursor at an entity
* Added: settings for a chat and for a gui
* Added: saving GUI chat data for a player
* Added: autohide gui for the mod
* Added: notices in gui
* Added: sending message in gui by enter
* Added: Ukrainian localization (localization by HAKER PLAY)
* Added: Spanish localization
* Added: Romanian localization (localization by anonymous#1)
* Added: Italian localization (localization by anonymous#1)
* Added: muting
* Changed: /toggle-chat drop-down
* Changed: style of GUI
* Changed: faction chat
* Changed: improved localization
* Changed: improved compatibility with scenarios, mods, soft-mods, game
* Changed: text field in gui not be cleared if message not be sended
* Changed: title
* Fixed: text in text field disappears when the interface was updated
* Code refactoring (not finished)

## 2018-01-14

### [v1.3.0][v1.3.0]

* Added: New GUI for chat
* Changed: command /toggle-chat [\<all/faction/allied/local/drop-down\>]
* Changed: (**temporary**) When you change the faction, depending on the presence of allies, the button "send to allies" **NOT** changes

## 2018-01-13

### [v1.2.4][v1.2.4]

* Changed: (local chat) Now the speed of the flying text is various from the length of the message

## 2018-01-07

### [v1.2.3][v1.2.3]

* Addded: raise event "on_console_chat"
* Changed: (local chat) If length message < 80 then create "flying text"
* Changed: locale for message in chat
* Fixed: the chat is not closed when using /toggle-chat
* Optimized code

### [v1.2.2][v1.2.2]

* Addded: When you change the faction, depending on the presence of allies, the button "send to allies" changes

### [v1.2.1][v1.2.1]

* Changed: description mod
* Changed: don't open allied chat, if there are no allies
* Fixed: variable "text_box" now local
* Optimized code

## 2018-01-06

### [v1.2.0][v1.2.0]

* Changed: command /toggle-chat [\<all/faction/allied/local\>] - disabled or re-enables your own parts of additional chat.

## 2018-01-04

### v1.1.2

* Added: commands /l \<text\> and /local-send \<text\> - sends a message to the nearest players

### v1.1.1

* Added: new flying text is slower and more time for life

### v1.1.0

* Added: local chat in gui
* Added: command /toggle-chat (disables or re-enables your own chat)

## 2017-12-20

### v1.0.0

* First release for 0.16

[v1.21.0]: https://mods.factorio.com/api/downloads/data/mods/2332/secondary-chat_1.21.0.zip
[v1.20.1]: https://mods.factorio.com/api/downloads/data/mods/2332/secondary-chat_1.20.1.zip
[v1.20.0]: https://mods.factorio.com/api/downloads/data/mods/2332/secondary-chat_1.20.0.zip
[v1.19.2]: https://mods.factorio.com/api/downloads/data/mods/2332/secondary-chat_1.19.2.zip
[v1.19.1]: https://mods.factorio.com/api/downloads/data/mods/2332/secondary-chat_1.19.1.zip
[v1.19.0]: https://mods.factorio.com/api/downloads/data/mods/2332/secondary-chat_1.19.0.zip
[v1.18.0]: https://mods.factorio.com/api/downloads/data/mods/2332/secondary-chat_1.18.0.zip
[v1.3.0]: https://mods.factorio.com/api/downloads/data/mods/2332/secondary-chat_1.3.0.zip
[v1.2.4]: https://mods.factorio.com/api/downloads/data/mods/2332/secondary-chat_1.2.4.zip
[v1.2.3]: https://mods.factorio.com/api/downloads/data/mods/2332/secondary-chat_1.2.3.zip
[v1.2.2]: https://mods.factorio.com/api/downloads/data/mods/2332/secondary-chat_1.2.2.zip
[v1.2.1]: https://mods.factorio.com/api/downloads/data/mods/2332/secondary-chat_1.2.1.zip
[v1.2.0]: https://mods.factorio.com/api/downloads/data/mods/2332/secondary-chat_1.2.0.zip


secondary-chat/control.lua
--[[
Copyright (C) 2017-2019 ZwerOxotnik <zweroxotnik@gmail.com>
Licensed under the EUPL, Version 1.2 only (the "LICENCE");
Author: ZwerOxotnik
Version: 1.21.0 (2019-04-08)

Description: Adds gui of chat, new commands, new types of chat, new interactions.

You can write and receive any information on the links below.
Source: https://gitlab.com/ZwerOxotnik/secondary-chat
Mod portal: https://mods.factorio.com/mod/secondary-chat
Homepage: https://forums.factorio.com/viewtopic.php?f=190&t=64625

]]--

local module = {}
module.version = "1.21.0"
module.events = {}
local BUILD = 1800 -- Always to increment this number when change the code

local get_event
if event_listener then
	get_event = function(name)
		return defines.events[name] or name
	end
else
	get_event = function(name)
		return defines.events[name]
	end
end

-- This function for compatibility with "Event listener" module and into other modules
local function put_event(event, func)
	event = get_event(event)
	if event then
		module.events[event] = func
		return true
	else
		log("That event is nil")
		-- error("That event is nil")
	end
	return false
end

chats = {}

max_autohide_time = 60 * 60 * 10 -- 10 min

require('secondary-chat/self_events')
color_picker = require('secondary-chat/integrations/color-picker')
require("secondary-chat/config/control")
require('secondary-chat/functions')
require('secondary-chat/gui/control')
require('secondary-chat/commands')
require('secondary-chat/chats/control')
if script.mod_name ~= 'level' then
	require('secondary-chat/mod-buttons')
end
require('secondary-chat/interface')

local function on_configuration_changed()
	update_global_config()

	for _, player in pairs( game.players ) do
		update_global_config_player(player)
	end
end

local function on_gui_click(event)
	-- Validation of data
	local gui = event.element
	if not (gui and gui.valid) then return end
	local player = game.players[event.player_index]
	if not (player and player.valid) then return end

	if (gui.name == "print_in_chat" or string.match(gui.name, "chat_(.+)")) then
		press_button_send_chat(event)
	else
		click_gui_chat(event)
	end
end

module.color_picker_ok_pressed = function(event)
	color_picker.ok_pressed(event)
end

local function on_player_removed(event)
	script.raise_event(chat_events.on_pre_delete_player_data, {player_index = event.player_index})

	global.secondary_chat.players[event.player_index] = nil
	global.secondary_chat.global.mutes[event.player_index] = nil

	update_chat_gui()
end

local function on_gui_text_changed(event)
	-- Validation of data
	local gui = event.element
	if not (gui and gui.valid) then return end
	local player = game.players[event.player_index]
	if not (player and player.valid) then return end

	if gui.name == 'chat_text_box' and gui.parent.parent.parent.name == 'table_chat' then
		if string.byte(gui.text, -1) == 10 then
			if #gui.text > 1 then
				gui.text = gui.text:sub(1, -2)
				event.element = gui.parent.parent.parent.select_chat.interactions.print_in_chat
				player_send_message(event)

				-- unfocus for the gui
				text_box = create_chat_text_box(gui.parent)

				if global.secondary_chat.players[event.player_index].settings.main.auto_focus.state then
					text_box.focus()
				end
			else
				-- unfocus for the gui
				create_chat_text_box(gui.parent)
			end
		end
		return true
	end
end

local function on_gui_checked_state_changed(event)
	-- Validation of data
	local gui = event.element
	if not (gui and gui.valid and gui.parent.parent.parent) then return end
	local player = game.players[event.player_index]
	if not (player and player.valid) then return end

	if gui.parent.name == 'config_table' and (gui.parent.parent.parent.name == 'settings' or gui.parent.parent.parent.parent.name == 'settings') then
		local parameter = string.match(gui.name, "(.+)_boolean")
		if parameter then
			update_checkbox(player, gui, parameter)
			return true
		end

		local parameter = string.match(gui.name, "(.+)-allow_fast_show")
		if parameter then
			update_allow_fast_show(player, gui, parameter)
			return true
		end
	end
end

local function on_gui_selection_state_changed(event)
	-- Validation of data
	local gui = event.element
	if not (gui and gui.valid) then return end
	local player = game.players[event.player_index]
	if not (player and player.valid) then return end

	if gui.parent.parent.name == 'select_chat' then
		if gui.parent.name == 'table_filter' then
			update_chat_and_drop_down(gui.parent.parent.interactions.chat_drop_down, player)
			return true
		elseif gui.name == 'chat_drop_down' then
			update_chat_and_drop_down(gui, player)
			return true
		elseif gui.name == 'targets_drop_down' then
			local chat_name = get_chat_name(gui.parent.parent.interactions.chat_drop_down.selected_index)
			script.raise_event(chat_events.on_update_chat_and_drop_down, {player_index = event.player_index, chat_name = chat_name, target = gui.items[gui.selected_index]})
			return true
		end
	end
end

local function on_player_promoted(event)
	-- Validation of data
	local player = game.players[event.player_index]
	if not (player and player.valid) then return end

	check_settings(player)
end

local function on_player_demoted(event)
	-- Validation of data
	local player = game.players[event.player_index]
	if not (player and player.valid) then return end

	check_settings(player)
end

module.on_load = function()
	chats = global.secondary_chat.chats
	if not game then
		if global.secondary_chat.state_chat == true then
			add_commands()
		end
	end

	local on_round_start = function()
		if not global.secondary_chat.state_chat then return end

		-- Create gui of chat
		for _, player in pairs( game.players ) do
			create_chat_gui(player)
		end
	end
	local on_round_end = function()
		if not global.secondary_chat.state_chat then return end

		-- Destroy gui of chat
		for _, player in pairs( game.players ) do
			destroy_chat_gui(player)
		end
	end

	-- Searching events "on_round_start" and "on_round_end"
	for interface, _ in pairs( remote.interfaces ) do
		local function_name = "get_event_name"
		if remote.interfaces[interface][function_name] then
			local ID_1 = remote.call(interface, function_name, "on_round_start")
			local ID_2 = remote.call(interface, function_name, "on_round_end")
			if (type(ID_1) == "number") and (type(ID_2) == "number") then
				if (script.get_event_handler(ID_1) == nil) and (script.get_event_handler(ID_2) == nil) then
					table.insert(module.events, ID_1, on_round_start)
					table.insert(module.events, ID_2, on_round_end)
					-- script.on_event(ID_1, on_round_start)
					-- script.on_event(ID_2, on_round_end)
				end
			end
		end
	end

    local function pick_interface(interfaces)
        for _, name in pairs( interfaces ) do
            if remote.interfaces[name] then
                return name
            end
        end

        return nil
    end

	-- Searching event "on_ok_button_clicked" from a mod "color-picker"
	local interface = pick_interface({"color-picker"})
	if interface then
		local ID_1 = remote.call(interface, "on_ok_button_clicked")
		if (type(ID_1) == "number" and script.get_event_handler(ID_1) == nil) then
			table.insert(module.events, ID_1, module.color_picker_ok_pressed)
			-- script.on_event(ID_1, module.color_picker_ok_pressed)
		end
	end
end

module.on_init = function()
	global_init()
	module.on_load()

	if global.secondary_chat.state_chat == true then
		add_commands()
	end
end

local function on_player_joined_game(event)
	if script.mod_name == 'level' and global.secondary_chat.build ~= BUILD then
		global_init()
		global.secondary_chat.build = BUILD
	end

	-- Validation of data
	local player = game.players[event.player_index]
	if not (player and player.valid) then return end

	if global.secondary_chat.players[event.player_index] then
		color_picker.destroy_gui(player)

		-- Remove settings
		local frame = player.gui.center.secondary_chat_settings
		if frame then
			frame.destroy()
		end

		local table_chat = player.gui.left.table_chat
		local settings = global.secondary_chat.players[event.player_index].settings
		if table_chat then
			table_chat.visible = settings.main.state_chat.state
			table_chat.top_chat.icons.color.visible = (global.secondary_chat.global.settings.main.allow_custom_color_message and (remote.interfaces["color-picker"] ~= nil))
		elseif settings.main.state_chat.state and not global.secondary_chat.state_chat then
			create_chat_gui(player)
		end
	else
		update_global_config_player(player)
	end

	update_chat_gui()
end

local function on_player_created()
	-- Delete previous chat GUI
	if #global.secondary_chat.players == 0 then
		for _, player in pairs( game.players ) do
			destroy_chat_gui(player)
		end
	end
end

local function on_player_left_game(event)
	-- Validation of data
	local player = game.players[event.player_index]
	if not (player and player.valid) then return end

	color_picker.destroy_gui(player)

	-- Hide chat
	local table_chat = player.gui.left.table_chat
	if table_chat then
		table_chat.visible = false
	end

	-- Remove settings
	local frame = player.gui.center.secondary_chat_settings
	if frame then
		frame.destroy()
	end

	global.secondary_chat.players[event.player_index].autohide = max_autohide_time
end

-- For soft-mods, scenarios, interfaces
module.delete = function(event)
	script.raise_event(chat_events.on_pre_remove_mod)

	remove_commands()
	for _, player in pairs( game.players ) do
		destroy_chat_gui(player)
		color_picker.destroy_gui(player)
	end
	remote.remove_interface('secondary-chat')
	global.secondary_chat = nil
end

local function on_player_muted(event)
	-- Validation of data
	local player = game.players[event.player_index]
	if not (player and player.valid) then return end

	-- Mute the player
	global.secondary_chat.global.mutes[event.player_index] = true
end

local function on_player_unmuted(event)
	-- Unmute the player
	global.secondary_chat.global.mutes[event.player_index] = nil
end


-- For attaching events
put_event("on_configuration_changed", on_configuration_changed)
put_event("on_gui_selection_state_changed", on_gui_selection_state_changed)
put_event("on_player_display_resolution_changed", check_settings_frame_size)
put_event("on_gui_checked_state_changed", on_gui_checked_state_changed)
put_event("on_gui_text_changed", on_gui_text_changed)
put_event("on_gui_click", on_gui_click)
put_event("on_player_created", on_player_created)
put_event("on_player_removed", on_player_removed)
put_event("on_player_joined_game", on_player_joined_game)
put_event("on_player_left_game", on_player_left_game)
put_event("on_player_changed_force", update_chat_gui)
put_event("on_forces_merging", update_chat_gui)
put_event("on_player_promoted", on_player_promoted)
put_event("on_player_demoted", on_player_demoted)
put_event("on_player_muted", on_player_muted)
put_event("on_player_unmuted", on_player_unmuted)

if script.mod_name ~= 'level' then
	module.custom_events =
	{
		autohide = function()
			local data = global.secondary_chat.players
			for player_index, player in pairs( game.connected_players ) do
				local table_chat = player.gui.left.table_chat
				if table_chat and table_chat.visible then
					if data[player_index].autohide <= 0 then
						table_chat.visible = false
						script.raise_event(chat_events.on_hide_gui_chat, {player_index = player_index, container = table_chat})
					else
						data[player_index].autohide = data[player_index].autohide - (60 * 60)
					end
				else
					data[player_index].autohide = max_autohide_time
				end
			end
		end
	}
end

return module
